lib: init ansi module

phip1611 opened the pull request:
The lib.ansi module provides a basic set of ANSI escape sequences to colorize and style text. The main purpose is to colorize trace messages (warnings, errors).

## Description of changes

Add `lib.ansi`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

phip1611 commented:
Ping @SuperSandro2000 @Lassulus

infinisil commented on lib/ansi.nix:
```diff
@@ -0,0 +1,42 @@
+    cyan = 6;
+    white = 7;
+  };
+  color = {
+    fg = builtins.mapAttrs (name: value: "${esc}[3${toString value}m") colorCode;
+    bg = builtins.mapAttrs (name: value: "${esc}[4${toString value}m") colorCode;
+  };
```
I think setting the background color is useless enough that we might not even want to have it, even just for completeness.

phip1611 replied:
I have no strong opinion, but please have a look at my comment below: https://github.com/NixOS/nixpkgs/pull/267189#issuecomment-1809660530

phip1611 replied:
Update: As written by @infinisil in https://github.com/NixOS/nixpkgs/pull/267189#issuecomment-1854687046

Functionality is/should be primarily for Nix, not for shell scripts.

infinisil commented on lib/ansi.nix:
```diff
@@ -0,0 +1,42 @@
+    fg = builtins.mapAttrs (name: value: "${esc}[3${toString value}m") colorCode;
+    bg = builtins.mapAttrs (name: value: "${esc}[4${toString value}m") colorCode;
+  };
+  stylize =
+    # List of ANSI style attributes.
+    styles:
+    # Text to encapsulate by style.
+    text:
```
This function would be better in a separate PR to keep the scope smaller.

infinisil reviewed (changes requested):
Generally I think it would be okay to have this, though I'd like to know more about what motivated this and the use cases.

Also this should have tests and documentation, see other parts of `lib` for reference.

phip1611 commented:
> Generally I think it would be okay to have this, though I'd like to know more about what motivated this and the use cases.

I'm thinking of two use cases:


# Styled trace messages

![image](https://github.com/NixOS/nixpkgs/assets/5737016/6038800a-6944-4adb-a2e5-b9f4edb59ff8)



# Styled shell scripts

![image](https://github.com/NixOS/nixpkgs/assets/5737016/a2b61511-7819-4756-a357-fe92232bc1a4)


For the latter, the best current alternative is to use `pkgs.ansi`:

![image](https://github.com/NixOS/nixpkgs/assets/5737016/bc015844-11a1-4e02-8c8f-aaa3ad2c2d3b)


I think, having `lib.ansi`, is a better approach. Limiting the functionality to basic colors and basic style attributes should be more than enough for now. I have no strong opinion regarding the background color, but it's a simple win and people might want to use it.

phip1611 committed:
lib: init ansi module

The lib.ansi module provides a basic set of ANSI escape sequences to
colorize and style text. This will improve usage of styled output from
Nix files (for warnings, errors, ...).

phip1611 committed:
lib: use ansi module in trivial.warn

phip1611 commented:
Any update, @infinisil @edolstra ?

SuperSandro2000 commented on lib/ansi.nix:
```diff
@@ -0,0 +1,59 @@
+    let
+      ansiBefore = builtins.concatStringsSep "" styles;
+    in
+    if text == ""
+    then ""
+    else "${ansiBefore}${text}${reset}";
```
I am not sure if the intended behavior to just concat the empty string would be what we want here

phip1611 replied:
Not sure what you mean. There is no concatenation of empty strings.

infinisil commented:
I'm not a fan of the use case for shell scripts. Nix is really not a general purpose language. It's slow, hard to test and hard to document. If something can be done with something other than Nix, that should be preferred.

If the bash code is too hard to read, you can also create functions to make it easier, just like Nix:
```bash
warnMsg() {
  echo -en "$(ansi yellow)$(ansi bold)WARN$(ansi reset): "
}

echo "$(warnMsg "WARN"): <some warning>"
```

You can even put that in a separate file or package for re-usability.

Alternatively you can use any other language than bash, which can have even more convenient ways of doing that.

infinisil commented:
I can see the use case for Nix error traces/messages though. There's no way around using Nix for that, and it could make the output nicer. I'm not sure if that's worth it though, I'm guessing that's not your main use case.

phip1611 commented:
Any update? @infinisil

phip1611 commented:
I rebased it and removed the background color support, @infinisil

phip1611 commented:
I rebased this to give it one more chance.
